### PR TITLE
update pattern type

### PR DESCRIPTION
**Pull request by shenjiangqiu**
the [pattern]( https://math.nist.gov/MatrixMarket/formats.html#MMformat) type is simply a zero-sized structure. update the io module to support:
1. read an MTX file with pattern type
2. read a normal non-pattern MTX file as a pattern-type MTX, ignore the data filed. #318

### CODE REVIEW

File: sprs/src/io.rs [hunk 1]
```diff
@@ -251,8 +260,19 @@ where
         if sym_mode == SymmetryMode::SkewSymmetric && row == col {
             return Err(BadMatrixMarketFile);
         }
-        if entry.next().is_some() {
-            return Err(BadMatrixMarketFile);
+        match droping_data {
```
Comment (mulimoen) [inline sprs/src/io.rs:263]: Replace `match` by `if`

File: sprs/src/io.rs [hunk 2]
```diff
@@ -802,4 +824,66 @@ mod test {
...
+    #[cfg_attr(miri, ignore)]
+    fn read_pattern_matrix() {
+        let path = "data/matrix_market/pattern.mm";
+        let mat = read_matrix_market::<Pattern, usize, _>(path).unwrap();
+        println!("trimat: {:?}", mat);
+        let csc = mat.to_csc();
+        assert_eq!(csc.get(0, 0), Some(&Pattern {}));
+        assert_eq!(csc.get(1, 1), Some(&Pattern {}));
+        assert_eq!(csc.get(2, 2), Some(&Pattern {}));
+        assert_eq!(csc.get(3, 3), Some(&Pattern {}));
```
Comment (mulimoen) [inline sprs/src/io.rs:838]: Replace `Pattern {}` by `Pattern` (drop braces)

File: sprs/src/num_kinds.rs [hunk 3]
```diff
@@ -2,13 +2,106 @@
...
+    fn add(self, _other: Pattern) -> Pattern {
+        Pattern {}
+    }
+}
+impl Zero for Pattern {
+    fn zero() -> Self {
+        Pattern {}
+    }
+    fn is_zero(&self) -> bool {
+        true
```
Comment (mulimoen) [inline sprs/src/num_kinds.rs:25]: This should be `false` as absence marks a zero

File: sprs/src/num_kinds.rs [hunk 4]
```diff
@@ -2,13 +2,106 @@
...
+
+impl Add for Pattern {
+    type Output = Pattern;
+    fn add(self, _other: Pattern) -> Pattern {
+        Pattern {}
+    }
+}
+impl Zero for Pattern {
+    fn zero() -> Self {
+        Pattern {}
```
Comment (mulimoen) [inline sprs/src/num_kinds.rs:22]: `Self`

File: sprs/src/num_kinds.rs [hunk 5]
```diff
@@ -2,13 +2,106 @@
...
+    }
+    fn is_zero(&self) -> bool {
+        true
+    }
+
+    fn set_zero(&mut self) {
+        *self = Zero::zero();
```
Comment (mulimoen) [inline sprs/src/num_kinds.rs:29]: Maybe panic instead?

File: sprs/src/num_kinds.rs [hunk 6]
```diff
@@ -2,13 +2,106 @@
...
+    }
+}
+impl Rem for Pattern {
```
Comment (mulimoen) [inline sprs/src/num_kinds.rs:32]: I am not sure why we need these traits, they make little sense for a pattern type

File: sprs/src/io.rs [hunk 7]
```diff
@@ -160,10 +163,16 @@ where
         DataType::Integer => NumKind::Integer,
         DataType::Real => NumKind::Float,
         DataType::Complex => NumKind::Complex,
+        DataType::Pattern => NumKind::Pattern,
     };
-    if N::num_kind() != data_num_kind {
+    // any type can be convert to pattern
```
Comment (mulimoen) [inline sprs/src/io.rs:168]: Do you intend for float matrices to be interpreted as a pattern matrix? I would prefer if all such conversions were done explicitly by the user by e.g. mapping after reading.

**Review by mulimoen — COMMENTED** [review]
Great to see pattern types being supported! Only some minor comments, this looks very promising.

File: sprs/src/num_kinds.rs (hunk 6 above)
Comment (shenjiangqiu) [reply sprs/src/num_kinds.rs]: Currently the to_csr and to_csc method needs <N> to be Num. Those traits are required to implement Num.
But in to_csr and to_csc, only Add and Zero are used. I suggest removing the Sum trait requirement for N. We can use N: Add<Output=N> instead. please check the newest Commit for detailed changes.

File: sprs/src/io.rs (hunk 7 above)
Comment (shenjiangqiu) [reply sprs/src/io.rs:165]: My consideration is that sometimes we have a lot of matrices in a folder, and we only care about the indices. Those matrices all have different types of data. It's hard for a user to read them all up and then convert them to the pattern.

File: sprs/src/num_kinds.rs (hunk 6 above)
Comment (shenjiangqiu) [reply sprs/src/num_kinds.rs]: The removal of the Num Trait .and then adding a new Add trait doesn't import break change. Cause all types implement Num must already implement Add.

File: sprs/src/num_kinds.rs (hunk 5 above)
Comment (shenjiangqiu) [reply sprs/src/num_kinds.rs]: I have deleted this trait.

File: sprs/src/num_kinds.rs (hunk 4 above)
Comment (shenjiangqiu) [reply sprs/src/num_kinds.rs]: I have deleted this trait.

File: sprs/src/num_kinds.rs (hunk 3 above)
Comment (shenjiangqiu) [reply sprs/src/num_kinds.rs]: I have deleted this trait.

File: sprs/src/io.rs (hunk 7 above)
Comment (shenjiangqiu) [reply sprs/src/io.rs:165]: because the user can't know the real type before an error is reported.

Comment (shenjiangqiu) [reply sprs/src/io.rs:165]: I have an idea. We can add a new function like `read_...._ignore_data(...)`, the old read function behavior like before(will return an error if the matrix is not originally Pattern), and the new `read_...._ignore_data(...)` always return CSMat<Pattern>, does this sound better?

**Pull request merged**